### PR TITLE
[trashbin] replace hook with storage wrapper

### PR DESCRIPTION
**Pull request by schiessle**
This is a minimal approach to replace the delete hook with a storage wrapper. This allows us to perform a move instead of a copy while moving files to the trash bin which will give us a huge performance improvement.

cc @DeepDiver1975 @icewind1991

### CODE REVIEW

**Comment by schiesbn** [comment]
@owncloud-bot retest this please

**Comment by DeepDiver1975** [comment]
@schiesbn 

local unit test execution helps ;-)

````
00:02:13.738 PHP Fatal error:  Class 'OCA\Files_Trashbin\OCP\Util' not found in /var/jenkins/workspace/pull-request-analyser-ng-simple@5/label/SLAVE/apps/files_trashbin/lib/trashbin.php on line 927
00:02:13.738 PHP Stack trace:
00:02:13.738 PHP   1. {main}() /usr/local/bin/phpunit:0
00:02:13.738 PHP   2. PHPUnit_TextUI_Command::main() /usr/local/bin/phpunit:612
00:02:13.738 PHP   3. PHPUnit_TextUI_Command->run() phar:///usr/local/bin/phpunit/phpunit/TextUI/Command.php:138
00:02:13.738 PHP   4. PHPUnit_TextUI_Command->handleArguments() phar:///usr/local/bin/phpunit/phpunit/TextUI/Command.php:148
00:02:13.738 PHP   5. PHPUnit_TextUI_Command->handleBootstrap() phar:///usr/local/bin/phpunit/phpunit/TextUI/Command.php:651
00:02:13.738 PHP   6. PHPUnit_Util_Fileloader::checkAndLoad() phar:///usr/local/bin/phpunit/phpunit/TextUI/Command.php:817
00:02:13.738 PHP   7. PHPUnit_Util_Fileloader::load() phar:///usr/local/bin/phpunit/phpunit/Util/Fileloader.php:77
00:02:13.738 PHP   8. include_once() phar:///usr/local/bin/phpunit/phpunit/Util/Fileloader.php:93
00:02:13.738 PHP   9. OC_App::loadApps() /var/jenkins/workspace/pull-request-analyser-ng-simple@5/label/SLAVE/tests/bootstrap.php:15
00:02:13.739 PHP  10. OC_App::loadApp() /var/jenkins/workspace/pull-request-analyser-ng-simple@5/label/SLAVE/lib/private/app.php:77
00:02:13.739 PHP  11. OC_App::requireAppFile() /var/jenkins/workspace/pull-request-analyser-ng-simple@5/label/SLAVE/lib/private/app.php:98
00:02:13.739 PHP  12. require_once() /var/jenkins/workspace/pull-request-analyser-ng-simple@5/label/SLAVE/lib/private/app.php:117
00:02:13.739 PHP  13. OCA\Files_Trashbin\Trashbin::registerHooks() /var/jenkins/workspace/pull-request-analyser-ng-simple@5/label/SLAVE/apps/files_trashbin/appinfo/app.php:5
00:02:13.986 Build step 'Execute shell' marked build as failure

**Comment by DeepDiver1975** [comment]
> local unit test execution helps ;-)

and using a proper ide :speak_no_evil:

File: apps/files_trashbin/lib/storage.php [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+namespace OCA\Files_Trashbin;
+
+use OC\Files\Storage\Wrapper\Wrapper;
+
+class Storage extends Wrapper {
+
+	private $mountPoint;
+	// remember already deleted files to avoid infinite loop if the trash bin
+	// move files across storages
+	private static $deletedFiles = array();
```
Comment (DeepDiver1975) [inline apps/files_trashbin/lib/storage.php]: does this have to be static? I mean: will there ever be files being deleted in two different storage implementations?

**Comment by MorrisJobke** [comment]
@schiesbn Can I ask you to label your PRs too? something like "4 - Developing" for WIP PRs or "5 - To review" for ready to review PRs. Then stuff like "app:files_trashbin" to label the component. This helps a lot to master the huge amount of PRs/issues and are just 3 clicks away ;)

**Comment by MorrisJobke** [comment]
```
There were 4 errors:
00:38:10.273 
00:38:10.273 1) Test_Files_Sharing_Mount::testMoveGroupShare
00:38:10.275 Invalid argument supplied for foreach()
00:38:10.275 
00:38:10.275 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/lib/private/files/cache/scanner.php:226
00:38:10.275 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/lib/private/files/cache/scanner.php:248
00:38:10.275 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/lib/private/files/cache/scanner.php:216
00:38:10.275 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/lib/private/files/view.php:930
00:38:10.275 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/lib/private/files/filesystem.php:782
00:38:10.275 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/apps/files_trashbin/lib/trashbin.php:39
00:38:10.275 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/apps/files_trashbin/lib/trashbin.php:145
00:38:10.275 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/apps/files_trashbin/lib/storage.php:48
00:38:10.275 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/lib/private/files/view.php:792
00:38:10.275 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/lib/private/files/view.php:437
00:38:10.275 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/apps/files_sharing/tests/sharedmount.php:45
00:38:10.276 
00:38:10.276 2) Test_Files_Sharing::testUnshareFromSelf
00:38:10.277 OCA\Files_Sharing\Exceptions\BrokenPath: Path does not start with /user/files
00:38:10.277 
00:38:10.277 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/apps/files_sharing/lib/sharedmount.php:103
00:38:10.277 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/apps/files_sharing/lib/sharedmount.php:121
00:38:10.277 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/lib/private/files/view.php:499
00:38:10.277 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/apps/files_trashbin/lib/trashbin.php:170
00:38:10.277 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/apps/files_trashbin/lib/storage.php:48
00:38:10.277 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/lib/private/files/view.php:792
00:38:10.277 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/lib/private/files/view.php:437
00:38:10.277 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/apps/files_sharing/tests/share.php:53
00:38:10.277 
00:38:10.277 3) Test_Files_Sharing::testShareAndUnshareFromSelf
00:38:10.278 OCA\Files_Sharing\Exceptions\BrokenPath: Path does not start with /user/files
00:38:10.278 
00:38:10.278 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/apps/files_sharing/lib/sharedmount.php:103
00:38:10.278 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/apps/files_sharing/lib/sharedmount.php:121
00:38:10.278 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/lib/private/files/view.php:499
00:38:10.278 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/apps/files_trashbin/lib/trashbin.php:170
00:38:10.278 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/apps/files_trashbin/lib/storage.php:48
00:38:10.278 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/lib/private/files/view.php:792
00:38:10.278 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/lib/private/files/view.php:437
00:38:10.278 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/apps/files_sharing/tests/share.php:53
00:38:10.278 
00:38:10.278 4) Test_Files_Sharing::testShareWithDifferentShareFolder
00:38:10.280 Exception: Sharing 5 failed, because this item is already shared with test-share-user2
00:38:10.280 
00:38:10.280 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/lib/private/share/share.php:597
00:38:10.280 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/lib/public/share.php:236
00:38:10.280 /var/jenkins/workspace/pull-request-analyser-ng-simple@2/label/SLAVE/apps/files_sharing/tests/share.php:231
```

File: apps/files_trashbin/lib/trashbin.php [hunk 2]
```diff
@@ -192,6 +194,8 @@ public static function move2trash($file_path) {
 			if ($user !== $owner) {
 				self::copyFilesToOwner($file_path, $owner, $ownerPath, $timestamp);
 			}
+
+			$result = true;
```
Comment (MorrisJobke) [inline apps/files_trashbin/lib/trashbin.php]: From scrutinizer:

The assignment to $result is dead and can be removed.

Comment (schiesbn) [inline apps/files_trashbin/lib/trashbin.php]: removed

**Comment by DeepDiver1975** [comment]
tested :+1:

**Comment by DeepDiver1975** [comment]
@icewind1991 @MorrisJobke @PVince81 can you please test this? THX

**Comment by Xenopathic** [comment]
Tested with/without external storages, :+1:

**Comment by MorrisJobke** [comment]
@schiesbn I got a question from @cdamken Is this the PR that fixes https://github.com/owncloud/client/issues/2858 and is it reasonable/possible to backport this to stable7 ?  cc @icewind1991

**Comment by DeepDiver1975** [comment]
> @schiesbn I got a question from @cdamken Is this the PR that fixes owncloud/client#2858 and is it reasonable/possible to backport this to stable7 ? cc @icewind1991

NO - additional fixes had to be applied on top - too risky - upgrade to 8 of disable trashbin

**Pull request merged**